internal/server/server.go

```diff
@@ -0,0 +1,280 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2022 Open Networking Foundation
+ */
+
+package server
+
+import (
+	"context"
+	"fmt"
+
+	pb "github.com/omec-project/pfcpsim/api"
+	"github.com/omec-project/pfcpsim/pkg/pfcpsim"
+	"github.com/omec-project/pfcpsim/pkg/pfcpsim/session"
+	ieLib "github.com/wmnsk/go-pfcp/ie"
+	"google.golang.org/grpc/codes"
+)
+
+// pfcpSimServer implements the Protobuf methods and keeps a connection to a remote PFCP Agent peer.
+// Its state is handled in internal/server/state.go
+type pfcpSimServer struct{}
+
+func (P pfcpSimServer) Recover(ctx context.Context, empty *pb.EmptyRequest) (*pb.Response, error) {
+
+	err := connectPFCPSim()
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.Response{
+		StatusCode: int32(codes.OK),
+		Message:    "Emulating Crash",
+	}, nil
+}
+
+func (P pfcpSimServer) Interrupt(ctx context.Context, empty *pb.EmptyRequest) (*pb.Response, error) {
+	sim.DisconnectN4()
+
+	return &pb.Response{
+		StatusCode: int32(codes.OK),
+		Message:    "Emulating Crash",
+	}, nil
+}
+
+func NewPFCPSimServer(remotePeerAddr string, upfAddr string, nodeBAddr string, ueAddrPool string) (*pfcpSimServer, error) {
+	var err error
+	localAddress, err = getLocalAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	remotePeerAddress = remotePeerAddr
+	upfAddress = upfAddr
+	nodeBAddress = nodeBAddr
+	ueAddressPool = ueAddrPool
+
+	// Connect internal PFCPSim to remote Peer
+	err = connectPFCPSim()
+	if err != nil {
+		return nil, err
+	}
+
+	return &pfcpSimServer{}, nil
+}
+
+func (P pfcpSimServer) Associate(ctx context.Context, empty *pb.EmptyRequest) (*pb.Response, error) {
+	err := sim.SetupAssociation()
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.Response{
+		StatusCode: int32(codes.OK),
+		Message:    "Association completed",
+	}, nil
```

Comment:
We should also log the result of the operation. Let me know if I can add the logging statement in this PR as it's already 'extensive'